### PR TITLE
CR-1126325 Parsing device_process.log and displaying end_of_simulatio…

### PR DESCRIPTION
**Pull request by ramyam-xilinx**
…n message (#6519)

* Parsing device_process.log generated from sim_ipc_master and display a message to exit on end_of_simulation

* Parsing device_process.log to check for end_of_simulation messages
(cherry picked from commit 38ad053493c5691d19a61cd404d72a950cc83061)

Conflicts:
	src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.h

### CODE REVIEW

File: src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx [hunk 1]
```diff
@@ -649,6 +651,11 @@ namespace xclcpuemhal2 {
 
     std::string binaryDirectory("");
     launchDeviceProcess(debuggable,binaryDirectory);
+    //Check if device_process.log already exists. Remove if exits.
+    std::string extIoTxtFile = deviceDirectory + "/../../../device_process.log";
+    FILE* fp = fopen(extIoTxtFile.c_str(), "rb");
```
Comment (stsoe) [inline src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx:656]: Where is this `fp` closed?

File: src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx [hunk 2]
```diff
@@ -1460,9 +1472,61 @@ namespace xclcpuemhal2 {
...
+   closemMessengerThread();
    saveDeviceProcessOutput();
   }
 
+  void CpuemShim::closemMessengerThread()
+  {
+    if (mMessengerThreadStarted) {
+      mMessengerThread.join();
+      mMessengerThreadStarted = false;
+    }
```
Comment (stsoe) [inline src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx:1484]: Will this not hang forever?   The thread loops on (inst) and I don't see it ever breaking out of the loop?

File: src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx [hunk 3]
```diff
@@ -1460,9 +1472,61 @@ namespace xclcpuemhal2 {
...
+      }
+    }
+  }
+  
+  int CpuemShim::parseLog()
+  {
+    std::vector<std::string> myvector = {"received request to end simulation from connected initiator"};
+    std::ifstream ifs(deviceDirectory + "/../../../device_process.log");
+	std::string logparse = deviceDirectory + "/../../../device_process.log";
+    if (ifs.is_open()) {
```
Comment (stsoe) [inline src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx:1512]: Use early exit, so revert the logic.

File: src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx [hunk 4]
```diff
@@ -1460,9 +1472,61 @@ namespace xclcpuemhal2 {
...
+      std::string line;
+      while (std::getline(ifs, line)) {
+        for (auto matchString : myvector) {
```
Comment (stsoe) [inline src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx:1515]: `const auto&`

File: src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx [hunk 5]
```diff
@@ -1460,9 +1472,61 @@ namespace xclcpuemhal2 {
...
+        if (parseCount%5 == 0 && timeCheck < 300) {
+          timeCheck += 10;
+        }
+      }
+    }
+  }
+  
+  int CpuemShim::parseLog()
+  {
+    std::vector<std::string> myvector = {"received request to end simulation from connected initiator"};
```
Comment (stsoe) [inline src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx:1509]: This vector has only one string?  I don't get why it is a vector?

**Review by stsoe — CHANGES_REQUESTED** [review]

**Review by stsoe — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx (hunk 2 above)
Comment (ramyam-xilinx) [reply src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx:1484]: We want the thread to be running forever (until emulation runs).Main intention being to parse and show some device process messages on the console. We are calling closemMessengerThread in the destructor and after xclClose call.

File: src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx [hunk 6]
```diff
@@ -1460,9 +1472,61 @@ namespace xclcpuemhal2 {
...
+      mMessengerThreadStarted = false;
+    }
+  }
+
+  void messagesThread(xclcpuemhal2::CpuemShim *inst)
+  {  
+    static auto start_time = std::chrono::high_resolution_clock::now();
+    unsigned int timeCheck = 0;
+    unsigned int parseCount = 0;
+    while (inst)
```
[the remark this answers is not in the document]
Comment (ramyam-xilinx) [reply src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx:1492]: We want the thread to be running forever (until emulation runs).Main intention being to parse and show some device process messages on the console. We are calling closemMessengerThread in the destructor and after xclClose call

File: src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx (hunk 5 above)
Comment (ramyam-xilinx) [reply src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx:1509]: We have a plan to parse other strings in future(based on the experience of hw emu driver).Hence used a vector.

File: src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx [hunk 7]
```diff
@@ -1223,9 +1233,61 @@ namespace xclcpuemhal2 {
...
+        if (parseCount%5 == 0 && timeCheck < 300) {
+          timeCheck += 10;
+        }
+      }
+    }
+  }
+  
+  int CpuemShim::parseLog()
+  {
+    std::vector<std::string> myvector = {"received request to end simulation from connected initiator"};
```
[the remark this answers is not in the document]
Comment (ramyam-xilinx) [reply src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx:1270]: These are similar changes for PCIE and Embedded sw emu driver . Currently , I donot find a place for common checkin. We will take it up as a future task.

**Comment by ramyam-xilinx** [comment]
The master branch PR for this change has been reverted. Hence closing this 2022.1 PR also. Once the changes are resubmitted and merged into the master branch we will resubmit to 2022.1 branch.

**Pull request closed without merging**